sawmill-core/src/main/java/io/logz/sawmill/processors/ExternalMappingSourceProcessor.java

```diff
@@ -0,0 +1,186 @@
+import io.logz.sawmill.exceptions.ProcessorConfigurationException;
+import io.logz.sawmill.http.ExternalMappingsClient;
+import io.logz.sawmill.utilities.JsonUtils;
+import java.net.MalformedURLException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+@ProcessorProvider(type = "externalMapping", factory = ExternalMappingSourceProcessor.Factory.class)
+public class ExternalMappingSourceProcessor implements Processor {
+
+    public static final int MINIMUM_REFRESH_PERIOD_IN_SECONDS = 15;
+    public static final int DISABLE_MAPPING_REFRESH = -1;
+
+    private final Logger logger = LoggerFactory.getLogger(ExternalMappingSourceProcessor.class);
+
+    private final String sourceField;
+    private final String targetField;
+    private final ExternalMappingsClient externalMappingsClient;
+    private final long mappingRefreshPeriodInSeconds;
+
+    private static ScheduledExecutorService scheduledRefreshMappingsExecutor;
+
+    @SuppressWarnings("Guava")
+    private final Supplier<Void> lazyInitSupplier;
+
+    private volatile Map<String, Iterable<String>> keyValueMappingsCache;
+
+    public ExternalMappingSourceProcessor(Configuration configuration) throws MalformedURLException {
+        this.sourceField = requireNonNull(configuration.getSourceField());
+        this.targetField = requireNonNull(configuration.getTargetField());
+        this.externalMappingsClient = new ExternalMappingsClient(configuration);
+        this.mappingRefreshPeriodInSeconds = configuration.getMappingRefreshPeriodInSeconds();
+
+        lazyInitSupplier = Suppliers.memoize(this::lazyInit);
+    }
+
+
+    private Void lazyInit() {
+        refreshExternalMapping();
+        if (mappingRefreshPeriodInSeconds != DISABLE_MAPPING_REFRESH) {
+            initThreadPool();
+            startScheduledExecutor();
+        }
+        /* return value is not used */
+        return null;
+    }
+
+    private void refreshExternalMapping() {
+        try {
+            keyValueMappingsCache = externalMappingsClient.loadMappings();
+        } catch (Exception e) {
+            logger.error("Cannot load external mapping for field {} due to an unexpected error", sourceField, e);
+        }
+
+        if (keyValueMappingsCache == null) {
+            keyValueMappingsCache = new HashMap<>();
+        }
+
+        if (keyValueMappingsCache.isEmpty()) {
+            logger.error("Cannot load external mapping for field: {}, received an empty map", sourceField);
+        }
+    }
+
+    private void initThreadPool() {
+        if (scheduledRefreshMappingsExecutor != null) return;
```

Comment:
@DanMelman easy, it's false only on the first time we call the `initThreadPool` method.
All, the subsequent invocation will return true.

<img width="1030" alt="image" src="https://user-images.githubusercontent.com/25157189/156542746-cb96a9f1-72d5-4ae0-b3ca-65bf11057f89.png">

<img width="1168" alt="image" src="https://user-images.githubusercontent.com/25157189/156542931-6c9fb7c7-a4c3-4ed7-abdc-923462fc826b.png">

Reply:
no, because `lazyInit()` is called only once per Processor instance due to the line 52 in constructor

```
lazyInitSupplier = Suppliers.memoize(this::lazyInit);
```

<img width="1368" alt="image" src="https://user-images.githubusercontent.com/25157189/156633478-71bb5db1-2c86-4309-a4ec-f1bfc56efff2.png">

https://guava.dev/releases/19.0/api/docs/com/google/common/base/Suppliers.html

ExternalMappingSourceProcessor is not singleton though.
Factory#create can be called multiple times, so this if-condition is added to prevent initializing multiple thread pools per Processor instance

Reply:
Got it!